File: mortality/julia/bench.jl
using MortalityTables
using BenchmarkTools

function runner()
    tbls = [MortalityTables.table(i) for i in 3299:3308];
    issue_ages = 18:50
    durations = 1:25
    timesteps = 0:29
    q = [
        tbl.select[issue_age][issue_age+(duration-1) + timestep]
        for timestep in timesteps, (tbl, issue_age, duration) in vec(collect(Iterators.product(tbls, issue_ages, durations)))
    ]
    npx = [1; cumprod(1 .- q[begin:end-1, :], dims=1)]
    v = 1/1.02
    v_eoy = v .^ (1 .+ timesteps) # this broadcasts
    unit_claims_discounted = npx .* q .* v_eoy
    sum(unit_claims_discounted)
end

@assert runner() == 1904.486552663679

b = @benchmark runner()

print("mean time: ", mean(b), "\nmedian time: ", median(b), "\nmemory: ", b.memory)

File: mortality/python/bench.py
from pymort.XML import MortXML
import numpy as np
import timeit


def runner():
    select = np.array(
        [MortXML(id).Tables[0].Values.unstack().values for id in range(3299, 3309)]
    )
    ultimate = np.array(
        [MortXML(id).Tables[1].Values.unstack().values for id in range(3299, 3309)]
    )

    mortality_table_index = np.arange(10)
    duration = np.arange(25)
    issue_age = np.arange(18, 51)
    mortality_table_index, duration, issue_age = [
        x.flatten() for x in np.meshgrid(mortality_table_index, duration, issue_age)
    ]
    time_axis = np.arange(30)[:, None]
    duration_projected = time_axis + duration

    q = np.where(
        duration_projected < select.shape[-1],
        select[
            mortality_table_index,
            issue_age - 18,
            np.minimum(duration_projected, select.shape[-1] - 1),
        ],  # np.minimum avoids some out of bounds error (which shouldn't exist in the first place? JAX works without workaround)
        ultimate[mortality_table_index, issue_age - 18 + duration_projected],
    )
    npx = np.concatenate(
        [np.ones((1, q.shape[1])), np.cumprod(1 - q, axis=0)[:-1]], axis=0
    )
    v = 1 / 1.02
    v_eoy = v ** np.arange(1, 31)[:, None]
    unit_claims_discounted = npx * q * v_eoy
    return np.sum(unit_claims_discounted)


assert runner() == 1904.4865526636793
res = timeit.timeit(stmt="runner()", setup="from __main__ import runner", number=20)
print("total time: ", res)
print("average time: ", res / 20)
